src/realm/sync/network/default_socket.cpp

```diff
@@ -491,70 +492,69 @@ void DefaultSocketProvider::start()
+    if (m_state == State::Starting || m_state == State::Running)
         return; // early return
 
     // If the thread has been previously run, make sure it has been joined first
-    if (m_state == State::Stopping || m_state == State::Stopped) {
-        if (m_thread.joinable()) {
-            lock.unlock();
-            m_thread.join();
-            lock.lock();
-        }
+    if (m_state == State::Stopping) {
+        state_wait_for(lock, State::Stopped);
     }
 
     m_logger_ptr->trace("Default event loop: start()");
     REALM_ASSERT(m_state == State::Stopped);
-    do_state_update(State::Starting);
+    do_state_update(lock, State::Starting);
     m_thread = std::thread{&DefaultSocketProvider::event_loop, this};
-    lock.unlock();
     // Wait for the thread to start before continuing
-    state_wait_for(State::Started);
+    state_wait_for(lock, State::Running);
 }
 
 void DefaultSocketProvider::event_loop()
 {
     m_logger_ptr->trace("Default event loop: thread running");
-    std::unique_lock<std::mutex> lock(m_mutex);
-    // The thread has started - don't return early due to stop() until after did_create_thread()
-    if (m_state == State::Starting) {
-        do_state_update(State::Started);
-    }
-    lock.unlock();
-
     auto will_destroy_thread = util::make_scope_exit([&]() noexcept {
         m_logger_ptr->trace("Default event loop: thread exiting");
         if (g_binding_callback_thread_observer)
             g_binding_callback_thread_observer->will_destroy_thread();
 
-        std::lock_guard<std::mutex> lock(m_mutex);
+        std::unique_lock<std::mutex> lock(m_mutex);
         // Did we get here due to an unhandled exception?
         if (m_state != State::Stopping) {
             m_logger_ptr->error("Default event loop: thread exited unexpectedly");
         }
-        do_state_update(State::Stopped);
+        m_state = State::Stopped;
+        std::notify_all_at_thread_exit(m_state_cv, std::move(lock));
     });
 
     if (g_binding_callback_thread_observer)
         g_binding_callback_thread_observer->did_create_thread();
 
-    lock.lock();
-    if (m_state != State::Started) {
-        // Stop has already been requested - exit early
-        lock.unlock(); // make sure the mutex is unloaded before exiting
-        return;        // early return
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        REALM_ASSERT(m_state == State::Starting);
     }
-    do_state_update(State::Running);
-    lock.unlock();
 
-    m_logger_ptr->trace("Default event loop: service run");
+    // We update the state to Running from inside the event loop so that start() is blocked until
+    // the event loop is actually ready to receive work.
+    m_service.post([this](Status status) {
+        std::unique_lock<std::mutex> lock(m_mutex);
+        if (status == ErrorCodes::OperationAborted) {
+            REALM_ASSERT(m_state == State::Stopping || m_state == State::Stopped);
```

Comment:
The only way I could think of for this to happen is if we created an event loop and then immediately destroyed it, so I don't think this is super likely to ever happen, so just REALM_ASSERT'ing this should be okay.

Reply:
The `ErrorCodes::OperationAborted` value for `Status` isn't implemented... it is always `Status::OK`
`network::Service` doesn't have a mechanism to abort anything on the queue when stop() is called. If there was nothing to run when stop() was called, it will exit immediately, but if there are items in the queue, it will continue processing them until it is empty.

Reply:
So is the right thing to do here to just return early if `m_state != Starting` ?

Reply:
We could potentially add support for OperationAborted, but I'm not sure if this will break anything...
```
# network.hpp:2638
inline void Service::post(H handler)
{
    // m_stopped is false when Service is created and after calling reset() after stop()
    if (m_impl->m_stopped) {
        handler(Status(ErrorCodes::OperationAborted), "Event loop is not running");
        return;
    }
    do_post(&Service::post_oper_constr<H>, sizeof(PostOper<H>), &handler);
}
```

Reply:
What happens to callbacks that have been posted, but not run, when the Service is destroyed?

Reply:
They just sit in the `m_completed_operations_2` queue and are destroyed when Service is destroyed - this means stale post function handlers (with potentially invalid objects) could be called when/if run() is called again. I think it would be preferable if this queue was flushed (or called with OperationAborted) when `reset()` is called.

I'll create a new task with this information and to be looked at separately (after platform networking).

Reply:
I've updated this function a bit so that if the service got reset (which I don't think it can be right now) and this callback ran, it has a generation count so that we only set the state to running if we're the correct call to run() for this event loop, and otherwise we return early if m_state == Stopping. I still check for status == OperationAborted, but just return early and don't assert anything.